<a href="https://colab.research.google.com/github/MeganKom/CS167Code/blob/main/notebook1_MeganKom.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook #1: Python, Pandas, and Pokemon, Oh My!

**Due 2/16 by 11:59pm**

The goal of this notebook is to become comfortable with using `python` and `pandas` as well as grabbing specific slices of the dataset. 

Start off by importing the data as we did in `practice-assignment`. Make sure the data is in your Google Drive, and that you update the path to point to where you have your data saved in Drive.

In [ ]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#import the data:

#make sure the path on the line below corresponds to the path where you put your dataset.
poke = pd.read_csv('/content/drive/MyDrive//pokemon.csv')
poke.head()

#                   Name Type 1  ... Speed  Generation  Legendary
0  1              Bulbasaur  Grass  ...    45           1      False
1  2                Ivysaur  Grass  ...    60           1      False
2  3               Venusaur  Grass  ...    80           1      False
3  3  VenusaurMega Venusaur  Grass  ...    80           1      False
4  4             Charmander   Fire  ...    65           1      False

[5 rows x 13 columns]

## Exercise #1: 
Print out all of the possible values for the primary type of pokemon.

In [ ]:
poke['Type 1'].unique()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

## Exercise #2:
Create a subset of data representing pokemon whose primary type is Bug 🐛, and the second type is Poison 💀. Save this DataFrame to the variable name `bug_poison_pokemon`. Display the first 5 lines of this subset.

In [ ]:
bug_poison_pokemon = poke[  (poke['Type 1'] == 'Bug') & (poke['Type 2'] == 'Poison')]
bug_poison_pokemon[:5]

#                   Name Type 1  ... Speed  Generation  Legendary
16  13                 Weedle    Bug  ...    50           1      False
17  14                 Kakuna    Bug  ...    35           1      False
18  15               Beedrill    Bug  ...    75           1      False
19  15  BeedrillMega Beedrill    Bug  ...   145           1      False
53  48                Venonat    Bug  ...    45           1      False

[5 rows x 13 columns]

## Exercise #3
Using `bug_poison_pokemon`, what is the median Attack of Bug Poison Pokemon?

In [ ]:
bug_poison_pokemon.groupby(['Attack']).median()
median_attack = bug_poison_pokemon['Attack'].median()
print(median_attack)

57.5


## Exercise #4
Which primary type of pokemon is the overall strongest? (not looking for philsophical debates here, I'm looking at the average of the 'Total' column, grouped by Type--so Fire v Water v Grass etc.).

In [ ]:
poke.groupby(['Type 1'])['Total'].mean()

Type 1
Bug         378.927536
Dark        445.741935
Dragon      550.531250
Electric    443.409091
Fairy       413.176471
Fighting    416.444444
Fire        458.076923
Flying      485.000000
Ghost       439.562500
Grass       421.142857
Ground      437.500000
Ice         433.458333
Normal      401.683673
Poison      399.142857
Psychic     475.947368
Rock        453.750000
Steel       487.703704
Water       430.455357
Name: Total, dtype: float64

**Based on the averages above, Dragon is the strongest at 550.53**

## Exercise #5
Choose a generation of pokemon, and create a subset with only pokemon with this generation. In this subset, only include the following 5 columns in this order: the Generation, name, type 1, type 2, and total.

In [ ]:
my_generation = poke[ (poke['Generation'] == 1)]
my_generation_display = my_generation[ ['Generation', 'Name', 'Type 1', 'Type 2', 'Total']]
print(my_generation_display)

     Generation                   Name   Type 1    Type 2  Total
0             1              Bulbasaur    Grass    Poison    318
1             1                Ivysaur    Grass    Poison    405
2             1               Venusaur    Grass    Poison    525
3             1  VenusaurMega Venusaur    Grass    Poison    625
4             1             Charmander     Fire       NaN    309
..          ...                    ...      ...       ...    ...
161           1              Dragonite   Dragon    Flying    600
162           1                 Mewtwo  Psychic       NaN    680
163           1    MewtwoMega Mewtwo X  Psychic  Fighting    780
164           1    MewtwoMega Mewtwo Y  Psychic       NaN    780
165           1                    Mew  Psychic       NaN    600

[166 rows x 5 columns]


## Exercise #6

From the Generation that you've chosen, ouput the mean, median, and standard deviation of the 'HP' column.

In [ ]:
generation_mean = my_generation['HP'].mean()
print('The mean is', generation_mean)

generation_median = my_generation['HP'].median()
print('The median is', generation_median)

generation_sd = my_generation['HP'].std()
print('The standard deviation is', generation_sd)

The mean is 65.81927710843374
The median is 62.0
The standard deviation is 28.15396826685769
